clang/include/clang/AST/PreorderAST.h

```diff
@@ -93,19 +93,15 @@ namespace clang {
     // that occur within nodes.
     // @return Returns a Lexicographic::Result indicating the comparison
     // between this and Other.
-    virtual Result Compare(const Node *Other, Lexicographic Lex) const {
-      return CompareKinds(Other);
-    }
+    virtual Result Compare(const Node *Other, Lexicographic Lex) const = 0;
 
     // Print the node.
     // @param[in] this is the current node of the AST.
-    virtual void PrettyPrint(llvm::raw_ostream &OS, ASTContext &Ctx) const { }
+    virtual void PrettyPrint(llvm::raw_ostream &OS, ASTContext &Ctx) const = 0;
 
     // Cleanup the memory consumed by this node.
     // @param[in] this is the current node of the AST.
-    virtual void Cleanup() {
-      delete this;
-    }
+    virtual void Cleanup() = 0;
```

Comment:
`Cleanup` should not be a pure virtual function otherwise the base class members like `Parent` won't be deleted. Please restore the previous definition of `Cleanup` in `Node`.